### ISSUE TITLE
`significant-drop-tightening` suggests syntax error `stdin.take(40).;`

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#15574]
Using the following flags

`--force-warn clippy::significant-drop-tightening`

this code:

```rust
use std::{io::{stdin, BufRead, Read}, process};
fn main() {
    //Let's read from stdin
    println!("Hello, what's your name?");
    let stdin = stdin().lock();
    let mut buffer = String::with_capacity(10);
    //Here we lock stdin and block to 10 bytes
    // Our string is now then only 10 bytes.
    //Even if it overflows like expected, it will reallocate.
    let mut stdin = stdin.take(40);
    if stdin.read_line(&mut buffer).is_err() {
        eprintln!("An error has occured while reading.");
        return;
    } //Now we print the result, our data is safe
    println!("Our string has a capacity of {}",buffer.capacity());
    println!("Hello {}!",buffer);
    //The string is freed automatically.
}
//Now let's try to overflow!! The string was reallocated.
```

caused the following diagnostics:

```
    Checking _snippet_0 v0.1.0 (/tmp/icemaker_global_tempdir.W9mV5T7suTWy/icemaker_clippyfix_tempdir.Z8u9dYrFB4Kw/_snippet_0)
warning: temporary with significant `Drop` can be early dropped
  --> src/main.rs:5:9
   |
 2 |   fn main() {
   |  ___________-
 3 | |     //Let's read from stdin
 4 | |     println!("Hello, what's your name?");
 5 | |     let stdin = stdin().lock();
   | |         ^^^^^
...  |
18 | | }
   | |_- temporary `stdin` is currently being dropped at the end of its contained scope
   |
   = note: this might lead to unnecessary resource contention
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#significant_drop_tightening
   = note: requested on the command line with `--force-warn clippy::significant-drop-tightening`
help: merge the temporary construction with its single usage
   |
 5 ~     
 6 +     let stdin = stdin().lock().take(40);
 7 |     let mut buffer = String::with_capacity(10);
...
10 |     //Even if it overflows like expected, it will reallocate.
11 ~     
   |

warning: temporary with significant `Drop` can be early dropped
  --> src/main.rs:10:13
   |
 2 |   fn main() {
   |  ___________-
 3 | |     //Let's read from stdin
 4 | |     println!("Hello, what's your name?");
 5 | |     let stdin = stdin().lock();
...  |
10 | |     let mut stdin = stdin.take(40);
   | |             ^^^^^
...  |
18 | | }
   | |_- temporary `stdin` is currently being dropped at the end of its contained scope
   |
   = note: this might lead to unnecessary resource contention
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#significant_drop_tightening
help: merge the temporary construction with its single usage
   |
10 ~     
11 +     stdin.take(40).;
12 ~      //Now we print the result, our data is safe
   |

warning: `_snippet_0` (bin "_snippet_0") generated 2 warnings
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.23s

```

However after applying these diagnostics, the resulting code:

```rust
use std::{io::{stdin, BufRead, Read}, process};
fn main() {
    //Let's read from stdin
    println!("Hello, what's your name?");
    let stdin = stdin().lock();
    let mut buffer = String::with_capacity(10);
    //Here we lock stdin and block to 10 bytes
    // Our string is now then only 10 bytes.
    //Even if it overflows like expected, it will reallocate.
    
    stdin.take(40).;
     //Now we print the result, our data is safe
    println!("Our string has a capacity of {}",buffer.capacity());
    println!("Hello {}!",buffer);
    //The string is freed automatically.
}
//Now let's try to overflow!! The string was reallocated.
```
no longer compiled:

```
    Checking _snippet_0 v0.1.0 (/tmp/icemaker_global_tempdir.W9mV5T7suTWy/icemaker_clippyfix_tempdir.Z8u9dYrFB4Kw/_snippet_0)
error: unexpected token: `;`
  --> src/main.rs:11:20
   |
11 |     stdin.take(40).;
   |                    ^

error: could not compile `_snippet_0` (bin "_snippet_0" test) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `_snippet_0` (bin "_snippet_0") due to 1 previous error

```

Version:
```
rustc 1.91.0-nightly (91ee6a405 2025-08-26)
binary: rustc
commit-hash: 91ee6a4057ce4bf1ab6d2f932cae497488d67c81
commit-date: 2025-08-26
host: x86_64-unknown-linux-gnu
release: 1.91.0-nightly
LLVM version: 21.1.0

```

**Comment by ada4a** [comment]
The architecture of the lint is quite complicated.. Here are the steps that ultimately result in this problem:
1. `StmtsChecker`, the main `Visitor` of the lint, has a field `ap: AuxParams`, which in turn has a field `apas: FxIndexMap<HirId, AuxParamsAttr>`, with `AuxParamsAttr` finally having a field `last_method_span: Span`, which is initialized to `DUMMY_SP`
2. In `StmtsChecker::visit_block`, for each statement of the block, we set `self.ap.curr_stmt` to, well, that statement, and then call `StmtChecker::visit_stmt` on that same statement
3. There, in one of the branches, we:
    1. Get a mutable reference to an existing `apa: AuxParamsAttr` from `self.ap.apas`
    2. Check `self.ap.curr_stmt.kind`, and, in cases `StmtKind::Let` and `StmtKind::Semi`, set `apa.last_method_span` to a non-dummy span
4. Later when emitting the lint, we construct the suggestion as
    ```
    {snippet(apa.last_stmt_span)}.{snippet(apa.last_method_span)};`
    ```

And that's where the problem arises.

1. In the reproducer you provided, the problematic `self.ap.curr_stmt` is the following statement:
    ```rs
    if stdin.read_line(&mut buffer).is_err() {
        eprintln!("An error has occured while reading.");
        return;
    }
    ```
2. Since its `kind` is `StmtKind::Expr` (because `if` doesn't have a trailing semicolon), we do not in fact update `apa.last_method_span`
3. Therefore, when constructing the suggestion, we have `snippet(apa.last_stmt_span) = "stdin.take(40)"` and `snippet(apa.last_method_span) = ""`, and so the suggestion comes out to be `stdin.take(40).;`

I quickly looked into handling the `StmtKind::Expr` case, but in the particular case of an if-statement, that's rather involved.

But I would argue the main problem here is using `DUMMY_SP` as a null value, so I came up with the following (interim) solution:
1. Change `apa.last_method_span` to be an `Option<Span>` instead
2. Currently, the lint uses `span_lint_and_then` to emit first a suggestion, then a note and label -- for now, I, rather crudely, abort the entire `span_lint_and_then` when `apa.last_method_span = None`:
```diff
diff --git a/clippy_lints/src/significant_drop_tightening.rs b/clippy_lints/src/significant_drop_tightening.rs
index 9110f684b..318099d9b 100644
--- a/clippy_lints/src/significant_drop_tightening.rs
+++ b/clippy_lints/src/significant_drop_tightening.rs
@@ -91,7 +91,11 @@ impl<'tcx> LateLintPass<'tcx> for SignificantDropTightening<'tcx> {
                         2 => {
                             let indent = " ".repeat(indent_of(cx, apa.last_stmt_span).unwrap_or(0));
                             let init_method = snippet(cx, apa.first_method_span, "..");
-                            let usage_method = snippet(cx, apa.last_method_span, "..");
+                            let usage_method = if let Some(last_method_span) = apa.last_method_span {
+                                snippet(cx, last_method_span, "..")
+                            } else {
+                                return;
+                            };
                             let stmt = if let Some(last_bind_ident) = apa.last_bind_ident {
                                 format!(
                                     "\n{indent}let {} = {init_method}.{usage_method};",
@@ -315,7 +319,7 @@ impl<'tcx> Visitor<'tcx> for StmtsChecker<'_, '_, '_, '_, 'tcx> {
                         if let Some(local_init) = local.init
                             && let hir::ExprKind::MethodCall(_, _, _, span) = local_init.kind
                         {
-                            apa.last_method_span = span;
+                            apa.last_method_span = Some(span);
                         }
                     },
                     hir::StmtKind::Semi(semi_expr) => {
@@ -325,7 +329,7 @@ impl<'tcx> Visitor<'tcx> for StmtsChecker<'_, '_, '_, '_, 'tcx> {
                             return;
                         }
                         if let hir::ExprKind::MethodCall(_, _, _, span) = semi_expr.kind {
-                            apa.last_method_span = span;
+                            apa.last_method_span = Some(span);
                         }
                     },
                     _ => {},
@@ -386,7 +390,7 @@ struct AuxParamsAttr {
     /// marked with `#[has_significant_drop]`.
     last_bind_ident: Option<Ident>,
     /// Similar to `last_bind_span` but encompasses the right-hand method call.
-    last_method_span: Span,
+    last_method_span: Option<Span>,
     /// Similar to `last_bind_span` but encompasses the whole contained statement.
     last_stmt_span: Span,
 }
@@ -402,7 +406,7 @@ impl Default for AuxParamsAttr {
             first_method_span: DUMMY_SP,
             first_stmt_span: DUMMY_SP,
             last_bind_ident: None,
-            last_method_span: DUMMY_SP,
+            last_method_span: None,
             last_stmt_span: DUMMY_SP,
         }
     }
```
I could of course modify it to only abort the _suggestion_ creation, leaving the note and label intact, since they're imo quite helpful even on their own. But other than that..
    
@c410-f3r, since you're the one who made the last big refactoring of the lint, I wanted to ask what you think of all this (I of course don't expect you to have retained much of the context, given that it's been 2 years).

**Comment by c410-f3r** [comment]
Well, I think any improvement is welcome as long as the current tests don't fail. Please feel free to provide any PRs that fix any issues.

IMO the current architecture shouldn't be taken too seriously because the lint still needs a lot of work that will eventually lead to more refactoring. For example, lifetimes aren't supported.

It's been 2 years but I still plan to work on this stuff. Hopefully `significant-drop-tightening` will become mature enough to move out of `nursery` in the near future.
